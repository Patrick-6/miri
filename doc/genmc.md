# **(WIP)** Documentation for Miri-GenMC
[GenMC](https://github.com/MPI-SWS/genmc) is a stateless model checker for exploring concurrent executions of a program.

**NOTE: Currently, no actual GenMC functionality is part of Miri, this is still WIP.**

<!-- FIXME(genmc): add explanation. -->

## Usage
Basic usage:
```shell
MIRIFLAGS="-Zmiri-genmc" cargo miri run
```

<!-- FIXME(genmc): explain options. -->

<!-- FIXME(genmc): explain Miri-GenMC specific functions. -->

## Tips

<!-- FIXME(genmc): add tips for using Miri-GenMC more efficiently. -->

## Limitations

Some or all of these limitations might get removed in the future:

- Borrow tracking is currently incompatible (stacked/tree borrows).
- Only Linux is supported for now.
- No 32-bit platform support.
- No cross-platform interpretation.

<!-- FIXME(genmc): document remaining limitations -->

## Development

GenMC is written in C++, which complicates development a bit.
For Rust-C++ interop, Miri uses [CXX.rs](https://cxx.rs/), and all handling of C++ code is contained in the `genmc-sys` crate (located in the Miri repository root directory: `miri/genmc-sys/`).

Building GenMC requires a compiler with C++23 support.
<!-- FIXME(genmc,llvm): remove once LLVM dependency is no longer required. -->
Currently, building GenMC also requires linking to LLVM, which needs to be installed manually.

The actual code for GenMC is not contained in the Miri repo itself, but in a [separate GenMC repo](https://github.com/MPI-SWS/genmc) (with different maintainers).
Note that this repo is just a mirror repo.
<!-- FIXME(genmc): define how submitting code to GenMC should be handled. -->

<!-- FIXME(genmc): explain development. -->

### Building the GenMC Library
The build script in the `genmc-sys` crate handles locating, downloading, building and linking the GenMC library.

To determine which GenMC repo path will be used, the following steps are taken:
- If the env var `GENMC_SRC_PATH` contains a path to a directory with a GenMC repo, use that path.
- If the path `genmc-sys/genmc` exists, use that path.
- If the path `genmc-sys/downloaded/genmc` exists, try to set the GenMC repo there to the commit we need.
- If the downloaded repo doesn't exist or is missing the commit, the build script will fetch the commit over the network.
  - Note that the build script will *not* access the network if any of the steps previous steps succeeds.

Once we get the path to the repo, the compilation proceeds in two steps:
- Compile GenMC into a library (using cmake).
- Compile the cxx.rs bridge to connect the library to the Rust code.
The first step is where all build settings are made, the relevant ones are then stored in a `config.h` file that can be included in the second compilation step.
