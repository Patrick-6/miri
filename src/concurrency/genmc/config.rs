use super::GenmcParams;

/// Configuration for GenMC mode.
/// The `params` field is shared with the C++ side.
/// The remaining options are kept on the Rust side.
#[derive(Debug, Default, Clone)]
pub struct GenmcConfig {
    pub(super) params: GenmcParams,
    do_estimation: bool,
    // FIXME(GenMC): add remaining options.
}

impl GenmcConfig {
    /// Function for parsing command line options for GenMC mode.
    ///
    /// All GenMC arguments start with the string "-Zmiri-genmc".
    /// Passing any GenMC argument will enable GenMC mode.
    ///
    /// `trimmed_arg` should be the argument to be parsed, with the suffix "-Zmiri-genmc" removed.
    pub fn parse_arg(genmc_config: &mut Option<GenmcConfig>, trimmed_arg: &str) {
        // FIXME(genmc,macos): Add `target_os = "macos"` once `https://github.com/dtolnay/cxx/issues/1535` is fixed.
        if !cfg!(all(
            feature = "genmc",
            any(target_os = "linux", target_os = "macos"),
            target_pointer_width = "64",
            target_endian = "little"
        )) {
            unimplemented!(
                "GenMC mode is not supported on this platform, cannot handle argument: \"-Zmiri-genmc{trimmed_arg}\""
            );
        }
        if genmc_config.is_none() {
            *genmc_config = Some(Default::default());
        }
        if trimmed_arg.is_empty() {
            return; // this corresponds to "-Zmiri-genmc"
        }
        // FIXME(GenMC): implement remaining parameters.
        todo!();
    }
}
