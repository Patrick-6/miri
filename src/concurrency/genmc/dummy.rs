use rustc_abi::{Align, Size};
use rustc_const_eval::interpret::{AllocId, InterpCx, InterpResult};
use rustc_middle::mir;

use crate::{
    AtomicFenceOrd, AtomicReadOrd, AtomicRwOrd, AtomicWriteOrd, MemoryKind, MiriConfig,
    MiriMachine, OpTy, Scalar, ThreadId, ThreadManager, VisitProvenance, VisitWith,
};

#[derive(Debug)]
pub struct GenmcCtx {}

#[derive(Debug, Default, Clone)]
pub struct GenmcConfig {}

// TODO GENMC: add all exposed methods here too

impl GenmcCtx {
    pub fn new(_miri_config: &MiriConfig, _genmc_config: &GenmcConfig) -> Self {
        unreachable!()
    }
    
    pub fn print_estimation_result(&self) {
        unreachable!()
    }

    pub fn get_blocked_execution_count(&self) -> usize {
        unreachable!()
    }

    pub fn print_genmc_graph(&self) {
        unreachable!()
    }

    pub fn is_exploration_done(&self) -> bool {
        unreachable!()
    }

    /**** Memory access handling ****/

    pub(crate) fn handle_execution_start(&self) {
        unreachable!()
    }

    pub(crate) fn handle_execution_end<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> Result<(), String> {
        unreachable!()
    }

    pub(super) fn set_ongoing_action_data_race_free(&self, _enable: bool) {
        unreachable!()
    }

    //* might fails if there's a race, load might also not read anything (returns None) */
    pub(crate) fn atomic_load<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicReadOrd,
        _old_val: Option<Scalar>,
    ) -> InterpResult<'tcx, Scalar> {
        unreachable!()
    }

    pub(crate) fn atomic_store<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _value: Scalar,
        _old_value: Option<Scalar>,
        _ordering: AtomicWriteOrd,
    ) -> InterpResult<'tcx, bool> {
        unreachable!()
    }

    pub(crate) fn atomic_fence<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _ordering: AtomicFenceOrd,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn atomic_rmw_op<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicRwOrd,
        (_rmw_op, _not): (mir::BinOp, bool),
        _rhs_scalar: Scalar,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, Scalar)> {
        unreachable!()
    }

    pub(crate) fn atomic_min_max_op<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicRwOrd,
        _min: bool,
        _is_signed: bool,
        _rhs_scalar: Scalar,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, Scalar)> {
        unreachable!()
    }

    pub(crate) fn atomic_exchange<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _rhs_scalar: Scalar,
        _ordering: AtomicRwOrd,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        unreachable!()
    }

    pub(crate) fn atomic_compare_exchange<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _expected_old_value: Scalar,
        _new_value: Scalar,
        _success: AtomicRwOrd,
        _fail: AtomicReadOrd,
        _can_fail_spuriously: bool,
        _old_value: Scalar,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        unreachable!()
    }

    pub(crate) fn memory_load<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _address: Size,
        _size: Size,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn memory_store<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _address: Size,
        _size: Size,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    /**** Memory (de)allocation ****/

    pub(crate) fn handle_alloc<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _alloc_id: AllocId,
        _size: Size,
        _alignment: Align,
        _memory_kind: MemoryKind,
    ) -> InterpResult<'tcx, u64> {
        unreachable!()
    }

    pub(crate) fn handle_dealloc<'tcx>(
        &self,
        _machine: &MiriMachine<'tcx>,
        _alloc_id: AllocId,
        _address: Size,
        _size: Size,
        _align: Align,
        _kind: MemoryKind,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    /**** Thread management ****/

    pub(crate) fn handle_thread_create<'tcx>(
        &self,
        _threads: &ThreadManager<'tcx>,
        _start_routine: crate::Pointer,
        _func_arg: &crate::ImmTy<'tcx>,
        _new_thread_id: ThreadId,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn handle_thread_join<'tcx>(
        &self,
        _active_thread_id: ThreadId,
        _child_thread_id: ThreadId,
    ) -> InterpResult<'tcx, ()> {
        unreachable!()
    }

    pub(crate) fn handle_thread_stack_empty<'tcx>(
        &self,
        _threads: &ThreadManager<'tcx>,
        _thread_id: ThreadId,
    ) {
        unreachable!()
    }

    pub(crate) fn handle_main_thread_stack_empty<'tcx>(&self, _threads: &ThreadManager<'tcx>) {
        unreachable!()
    }

    pub(crate) fn handle_thread_finish<'tcx>(&self, _threads: &ThreadManager<'tcx>) {
        unreachable!()
    }

    /**** Scheduling functionality ****/

    pub(crate) fn schedule_thread<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> InterpResult<'tcx, ThreadId> {
        unreachable!()
    }
}

/// Other functionality not directly related to event handling
impl<'tcx> EvalContextExt<'tcx> for crate::MiriInterpCx<'tcx> {}
pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
    fn check_genmc_intercept_function(
        &mut self,
        _instance: rustc_middle::ty::Instance<'tcx>,
        _args: &[rustc_const_eval::interpret::FnArg<'tcx, crate::Provenance>],
        _dest: &crate::PlaceTy<'tcx>,
        _ret: Option<mir::BasicBlock>,
    ) -> InterpResult<'tcx, bool> {
        unreachable!()
    }

    /**** Blocking instructions ****/

    fn handle_genmc_verifier_assume(&mut self, _condition: &OpTy<'tcx>) -> InterpResult<'tcx> {
        unreachable!()
    }
}

impl VisitProvenance for GenmcCtx {
    fn visit_provenance(&self, _visit: &mut VisitWith<'_>) {
        unreachable!()
    }
}

impl GenmcConfig {
    pub fn parse_arg(_genmc_config: &mut Option<GenmcConfig>, trimmed_arg: &str) {
        unimplemented!(
            "GenMC feature im Miri is disabled, cannot handle argument: \"-Zmiri-genmc{trimmed_arg}\""
        );
    }

    pub fn print_exec_graphs(&self) -> bool {
        unreachable!()
    }

    pub fn do_estimation(&self) -> bool {
        unreachable!()
    }
}
