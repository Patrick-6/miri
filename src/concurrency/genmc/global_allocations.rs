use std::cmp::max;
use std::collections::hash_map::Entry;
use std::sync::RwLock;

use genmc_sys::{GENMC_GLOBAL_ADDRESSES_MASK, getGlobalAllocStaticMask};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use rustc_const_eval::interpret::{
    AllocId, AllocInfo, AllocKind, InterpResult, PointerArithmetic, interp_ok,
};
use rustc_data_structures::fx::FxHashMap;
use rustc_middle::{err_exhaust, throw_exhaust};
use tracing::info;

use crate::alloc_addresses::align_addr;

#[derive(Debug, Default)]
pub struct GlobalAllocationHandler {
    inner: RwLock<GlobalStateInner>,
}

/// This contains more or less a subset of the functionality of `struct GlobalStateInner` in `alloc_addresses`.
#[derive(Clone, Debug)]
struct GlobalStateInner {
    /// This is used as a map between the address of each allocation and its `AllocId`. It is always
    /// sorted by address. We cannot use a `HashMap` since we can be given an address that is offset
    /// from the base address, and we need to find the `AllocId` it belongs to. This is not the
    /// *full* inverse of `base_addr`; dead allocations have been removed.
    #[allow(unused)] // FIXME(GenMC): do we need this?
    int_to_ptr_map: Vec<(u64, AllocId)>,
    /// The base address for each allocation.
    /// This is the inverse of `int_to_ptr_map`.
    base_addr: FxHashMap<AllocId, u64>,
    /// This is used as a memory address when a new pointer is casted to an integer. It
    /// is always larger than any address that was previously made part of a block.
    next_base_addr: u64,
    /// To add some randomness to the allocations
    /// FIXME(GenMC): maybe seed this from the rng in MiriMachine?
    rng: StdRng,
}

impl Default for GlobalStateInner {
    fn default() -> Self {
        Self::new()
    }
}

impl GlobalStateInner {
    pub fn new() -> Self {
        assert_eq!(GENMC_GLOBAL_ADDRESSES_MASK, getGlobalAllocStaticMask());
        assert_ne!(GENMC_GLOBAL_ADDRESSES_MASK, 0);
        Self {
            int_to_ptr_map: Vec::default(),
            base_addr: FxHashMap::default(),
            next_base_addr: GENMC_GLOBAL_ADDRESSES_MASK,
            rng: StdRng::seed_from_u64(0),
        }
    }

    fn global_allocate_addr<'tcx>(
        &mut self,
        alloc_id: AllocId,
        info: AllocInfo,
    ) -> InterpResult<'tcx, u64> {
        let entry = match self.base_addr.entry(alloc_id) {
            Entry::Occupied(occupied_entry) => {
                // Looks like some other thread allocated this for us
                // between when we released the read lock and aquired the write lock,
                // so we just return that value.
                return interp_ok(*occupied_entry.get());
            }
            Entry::Vacant(vacant_entry) => vacant_entry,
        };

        // This is either called immediately after allocation (and then cached), or when
        // adjusting `tcx` pointers (which never get freed). So assert that we are looking
        // at a live allocation. This also ensures that we never re-assign an address to an
        // allocation that previously had an address, but then was freed and the address
        // information was removed.
        assert!(!matches!(info.kind, AllocKind::Dead));

        // This allocation does not have a base address yet, pick or reuse one.

        // We are not in native lib mode, so we control the addresses ourselves.

        // We have to pick a fresh address.
        // Leave some space to the previous allocation, to give it some chance to be less aligned.
        // We ensure that `(global_state.next_base_addr + slack) % 16` is uniformly distributed.
        let slack = self.rng.random_range(0..16);
        // From next_base_addr + slack, round up to adjust for alignment.
        let base_addr =
            self.next_base_addr.checked_add(slack).ok_or_else(|| err_exhaust!(AddressSpaceFull))?;
        let base_addr = align_addr(base_addr, info.align.bytes());

        // Remember next base address.  If this allocation is zero-sized, leave a gap of at
        // least 1 to avoid two allocations having the same base address. (The logic in
        // `alloc_id_from_addr` assumes unique addresses, and different function/vtable pointers
        // need to be distinguishable!)
        self.next_base_addr = base_addr
            .checked_add(max(info.size.bytes(), 1))
            .ok_or_else(|| err_exhaust!(AddressSpaceFull))?;

        assert_ne!(0, base_addr & GENMC_GLOBAL_ADDRESSES_MASK);
        assert_ne!(0, self.next_base_addr & GENMC_GLOBAL_ADDRESSES_MASK);
        // Cache the address for future use.
        entry.insert(base_addr);

        interp_ok(base_addr)
    }
}

// FIXME(GenMC): "ExtPriv" or "PrivExt"?
impl<'tcx> EvalContextExtPriv<'tcx> for crate::MiriInterpCx<'tcx> {}
pub(super) trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
    fn get_global_allocation_address(
        &self,
        global_allocation_handler: &GlobalAllocationHandler,
        alloc_id: AllocId,
    ) -> InterpResult<'tcx, u64> {
        let this = self.eval_context_ref();
        let info = this.get_alloc_info(alloc_id);

        let global_state = global_allocation_handler.inner.read().unwrap();
        if let Some(base_addr) = global_state.base_addr.get(&alloc_id) {
            info!(
                "GenMC: address for global with alloc id {alloc_id:?} was cached: {base_addr} == {base_addr:#x}"
            );
            return interp_ok(*base_addr);
        }

        drop(global_state);
        // We need to upgrade to a write lock. std::sync::RwLock doesn't support this, so we drop the guard and lock again
        // Note that another thread might run in between and allocate the address, but we handle this case in the allocation function.
        let mut global_state = global_allocation_handler.inner.write().unwrap();
        let base_addr = global_state.global_allocate_addr(alloc_id, info)?;
        // Even if `Size` didn't overflow, we might still have filled up the address space.
        if global_state.next_base_addr > this.target_usize_max() {
            throw_exhaust!(AddressSpaceFull);
        }
        info!(
            "GenMC: global with alloc id {alloc_id:?} got address: {base_addr} == {base_addr:#x}"
        );
        interp_ok(base_addr)
    }
}
