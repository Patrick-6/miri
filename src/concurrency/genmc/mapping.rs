use genmc_sys::{MemOrdering, RMWBinOp};

use crate::{AtomicFenceOrd, AtomicReadOrd, AtomicRwOrd, AtomicWriteOrd};

impl AtomicReadOrd {
    pub(super) fn convert(self) -> MemOrdering {
        match self {
            AtomicReadOrd::Relaxed => MemOrdering::Relaxed,
            AtomicReadOrd::Acquire => MemOrdering::Acquire,
            AtomicReadOrd::SeqCst => MemOrdering::SequentiallyConsistent,
        }
    }
}

impl AtomicWriteOrd {
    pub(super) fn convert(self) -> MemOrdering {
        match self {
            AtomicWriteOrd::Relaxed => MemOrdering::Relaxed,
            AtomicWriteOrd::Release => MemOrdering::Release,
            AtomicWriteOrd::SeqCst => MemOrdering::SequentiallyConsistent,
        }
    }
}

impl AtomicFenceOrd {
    pub(super) fn convert(self) -> MemOrdering {
        match self {
            AtomicFenceOrd::Acquire => MemOrdering::Acquire,
            AtomicFenceOrd::Release => MemOrdering::Release,
            AtomicFenceOrd::AcqRel => MemOrdering::AcquireRelease,
            AtomicFenceOrd::SeqCst => MemOrdering::SequentiallyConsistent,
        }
    }
}

impl AtomicRwOrd {
    /// Split up an atomic read-write memory ordering into a separate read and write ordering.
    pub(super) fn split_memory_orderings(self) -> (AtomicReadOrd, AtomicWriteOrd) {
        match self {
            AtomicRwOrd::Relaxed => (AtomicReadOrd::Relaxed, AtomicWriteOrd::Relaxed),
            AtomicRwOrd::Acquire => (AtomicReadOrd::Acquire, AtomicWriteOrd::Relaxed),
            AtomicRwOrd::Release => (AtomicReadOrd::Relaxed, AtomicWriteOrd::Release),
            AtomicRwOrd::AcqRel => (AtomicReadOrd::Acquire, AtomicWriteOrd::Release),
            AtomicRwOrd::SeqCst => (AtomicReadOrd::SeqCst, AtomicWriteOrd::SeqCst),
        }
    }

    /// Split up the atomic success ordering of a read-modify-write operation into GenMC's representation.
    /// Note that both returned orderings are currently identical, because this is what GenMC expects.
    pub(super) fn to_genmc_memory_orderings(self) -> (MemOrdering, MemOrdering) {
        match self {
            AtomicRwOrd::Relaxed => (MemOrdering::Relaxed, MemOrdering::Relaxed),
            AtomicRwOrd::Acquire => (MemOrdering::Acquire, MemOrdering::Acquire),
            AtomicRwOrd::Release => (MemOrdering::Release, MemOrdering::Release),
            AtomicRwOrd::AcqRel => (MemOrdering::AcquireRelease, MemOrdering::AcquireRelease),
            AtomicRwOrd::SeqCst =>
                (MemOrdering::SequentiallyConsistent, MemOrdering::SequentiallyConsistent),
        }
    }
}

pub(super) fn min_max_to_genmc_rmw_op(min: bool, is_signed: bool) -> RMWBinOp {
    match (min, is_signed) {
        (true, true) => RMWBinOp::Min, // TODO GENMC: is there a use for FMin? (Min, UMin, FMin)
        (false, true) => RMWBinOp::Max,
        (true, false) => RMWBinOp::UMin,
        (false, false) => RMWBinOp::UMax,
    }
}

pub(super) fn to_genmc_rmw_op(bin_op: rustc_middle::mir::BinOp, negate: bool) -> RMWBinOp {
    match bin_op {
        rustc_middle::mir::BinOp::Add => RMWBinOp::Add,
        rustc_middle::mir::BinOp::Sub => RMWBinOp::Sub,
        rustc_middle::mir::BinOp::BitOr if !negate => RMWBinOp::Or,
        rustc_middle::mir::BinOp::BitXor if !negate => RMWBinOp::Xor,
        rustc_middle::mir::BinOp::BitAnd if negate => RMWBinOp::Nand,
        rustc_middle::mir::BinOp::BitAnd => RMWBinOp::And,
        _ => {
            panic!("unsupported atomic operation: bin_op: {bin_op:?}, negate: {negate}");
        }
    }
}
