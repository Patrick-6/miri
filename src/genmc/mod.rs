use std::cell::{Cell, RefCell};

use rustc_abi::{Align, Size};
use rustc_const_eval::interpret::{InterpCx, InterpResult, interp_ok};

use self::helper::{
    NextInstrInfo, Threads, genmc_scalar_to_scalar, get_next_instr_info,
    rhs_scalar_to_genmc_scalar, scalar_to_genmc_scalar,
};
use self::thread_info_manager::{GenmcThreadId, GenmcThreadIdInner, ThreadInfoManager};
use crate::intrinsics::AtomicOp;
use crate::{
    AtomicFenceOrd, AtomicReadOrd, AtomicRwOrd, AtomicWriteOrd, MemoryKind, MiriConfig,
    MiriMachine, Scalar, TerminationInfo, ThreadId, ThreadManager, VisitProvenance, VisitWith,
};

mod config;

pub use self::config::GenmcConfig;
pub use self::ffi::GenmcParams;

// TODO: add fields
pub struct GenmcCtx {
    /// Some actions Miri does are allowed to cause data races.
    /// GenMC will not be informed about certain actions (e.g. non-atomic loads) when this flag is set
    allow_data_races: Cell<bool>,
}

impl GenmcCtx {
    /// Validate the selected configuration options and create a new `GenmcCtx` if successful
    ///
    /// Some combinations of options are (currently) not allowed:
    /// - Aliasing model checking is incompatible with GenMC mode
    ///   - The reason is that the required information is lost when pointers are send to GenMC and back
    /// - Data race checking and weak memory emulation must be turned off, since GenMC does this by itself
    /// - "Many seeds" mode in Miri is currently incompatible with GenMC mode
    pub fn try_new(_miri_config: &MiriConfig) -> Option<Self> {
        todo!()
    }

    pub fn get_stuck_execution_count(&self) -> usize {
        todo!()
    }

    pub fn print_genmc_graph(&self) {
        todo!()
    }

    pub fn is_exploration_done(&self) -> bool {
        todo!()
    }

    pub(crate) fn handle_execution_start(&self) {
        todo!()
    }

    pub(crate) fn handle_execution_end<'tcx>(
        &self,
        _thread_manager: &ThreadManager<'tcx>,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> Result<(), String> {
        todo!()
    }

    /**** Memory access handling ****/

    /// If `true` is passed, allow for data races to happen without triggering an error, until this function is called again with argument `false`.
    /// This allows for racy non-atomic memory accesses to be ignored (GenMC is not informed about them at all).
    ///
    /// Certain operations are not permitted in GenMC mode with data races disabled and will cause a panic, e.g., atomic accesses or asking for scheduling decisions.
    ///
    /// # Panics
    /// This method will panic if data races are nested
    pub(crate) fn set_ongoing_action_data_race_free(&self, enable: bool) {
        let old = self.allow_data_races.replace(enable);
        assert_ne!(old, enable, "cannot nest allow_data_races");
    }

    pub(crate) fn atomic_load<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _ordering: AtomicReadOrd,
        _old_val: Option<Scalar>,
    ) -> InterpResult<'tcx, Scalar> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_store<'tcx>(
        &self,
        _ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        _address: Size,
        _size: Size,
        _value: Scalar,
        _ordering: AtomicWriteOrd,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_fence<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        ordering: AtomicFenceOrd,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_rmw_op<'tcx>(
        &self,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        address: Size,
        size: Size,
        ordering: AtomicRwOrd,
        rmw_op: AtomicOp,
        rhs_scalar: Scalar,
        is_unsigned: bool,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_exchange<'tcx>(
        &self,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        address: Size,
        size: Size,
        rhs_scalar: Scalar,
        ordering: AtomicRwOrd,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn atomic_compare_exchange<'tcx>(
        &self,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
        address: Size,
        size: Size,
        expected_old_value: Scalar,
        new_value: Scalar,
        success: AtomicRwOrd,
        fail: AtomicReadOrd,
        can_fail_spuriously: bool,
    ) -> InterpResult<'tcx, (Scalar, bool)> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn memory_load<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        address: Size,
        size: Size,
    ) -> InterpResult<'tcx, Scalar> {
        todo!()
    }

    pub(crate) fn memory_store<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        address: Size,
        size: Size,
    ) -> InterpResult<'tcx, ()> {
        todo!()
    }

    /**** Memory (de)allocation ****/

    pub(crate) fn handle_alloc<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        size: Size,
        alignment: Align,
        memory_kind: MemoryKind,
    ) -> InterpResult<'tcx, u64> {
        todo!()
    }

    pub(crate) fn handle_dealloc<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        address: Size,
        size: Size,
        align: Align,
        kind: MemoryKind,
    ) -> InterpResult<'tcx, ()> {
        todo!()
    }

    /**** Thread management ****/

    pub(crate) fn handle_thread_create<'tcx>(
        &self,
        threads: &ThreadManager<'tcx>,
        new_thread_id: ThreadId,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn handle_thread_join(
        &self,
        active_thread_id: ThreadId,
        child_thread_id: ThreadId,
    ) -> Result<(), ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    pub(crate) fn handle_thread_stack_empty(&self, _thread_id: ThreadId) {
        todo!()
    }

    pub(crate) fn handle_thread_finish<'tcx>(
        &self,
        threads: &ThreadManager<'tcx>,
    ) -> InterpResult<'tcx, ()> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    /**** Scheduling functionality ****/

    pub(crate) fn should_preempt(&self) -> bool {
        true // TODO GENMC
    }

    pub(crate) fn schedule_thread<'tcx>(
        &self,
        thread_manager: &ThreadManager<'tcx>,
        ecx: &InterpCx<'tcx, MiriMachine<'tcx>>,
    ) -> InterpResult<'tcx, ThreadId> {
        assert!(!self.allow_data_races.get());
        todo!()
    }

    /**** Blocking instructions ****/

    pub(crate) fn handle_verifier_assume<'tcx>(
        &self,
        machine: &MiriMachine<'tcx>,
        condition: bool,
    ) -> InterpResult<'tcx, ()> {
        if condition { interp_ok(()) } else { self.handle_user_block(machine) }
    }
}

impl VisitProvenance for GenmcCtx {
    fn visit_provenance(&self, _visit: &mut VisitWith<'_>) {
        // We don't have any tags.
    }
}

impl GenmcCtx {
    fn handle_user_block<'tcx>(&self, machine: &MiriMachine<'tcx>) -> InterpResult<'tcx, ()> {
        todo!()
    }
}
