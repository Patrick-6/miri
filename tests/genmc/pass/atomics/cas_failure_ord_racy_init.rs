//@compile-flags: -Zmiri-genmc -Zmiri-disable-stacked-borrows -Zmiri-genmc-verbose
//@normalize-stderr-test: "Verification took .*s" -> "Verification took [TIME]s"

#![no_main]

#[path = "../../../utils/genmc.rs"]
mod genmc;
#[path = "../../../utils/mod.rs"]
mod utils;

use std::sync::atomic::AtomicU64;
use std::sync::atomic::Ordering::*;

use crate::genmc::*;
use crate::utils::*;

const UNINITIALIZED: u64 = 0;
const GETTING_INIT: u64 = 1;
const INITIALIZED: u64 = 2;

static mut X: u64 = 0;
static LOCK: AtomicU64 = AtomicU64::new(UNINITIALIZED);

#[unsafe(no_mangle)]
fn miri_start(_argc: isize, _argv: *const *const u8) -> isize {
    unsafe {
        spawn_pthread_closure(|| {
            get_or_init(1);
        });
        spawn_pthread_closure(|| {
            get_or_init(2);
        });
    }
    0
}

fn get_or_init(value: u64) -> u64 {
    loop {
        match LOCK.load(Acquire) {
            INITIALIZED => unsafe { return X },
            GETTING_INIT => {
                unsafe { miri_genmc_assume(false) };
                continue;
            }
            UNINITIALIZED => {
                if LOCK
                    .compare_exchange_weak(UNINITIALIZED, GETTING_INIT, Relaxed, Relaxed)
                    .is_err()
                {
                    unsafe { miri_genmc_assume(false) };
                    continue;
                }
                unsafe { X = value };
                if LOCK.compare_exchange(GETTING_INIT, INITIALIZED, Release, Relaxed).is_err() {
                    unsafe { std::hint::unreachable_unchecked() };
                }
                return value;
            }
            _ => unsafe { std::hint::unreachable_unchecked() },
        }
    }
}
