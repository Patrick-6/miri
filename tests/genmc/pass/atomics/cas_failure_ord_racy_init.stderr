Running GenMC Verification...
warning: GenMC currently does not model spurious failures of `compare_exchange_weak`. Miri with GenMC might miss bugs related to spurious failures.
  --> tests/genmc/pass/atomics/cas_failure_ord_racy_init.rs:LL:CC
   |
LL |                   if LOCK
   |  ____________________^
LL | |                     .compare_exchange_weak(UNINITIALIZED, GETTING_INIT, Relaxed, Relaxed)
   | |_________________________________________________________________________________________^ GenMC might miss possible behaviors of this code
   |
   = note: BACKTRACE on thread `unnamed-ID`:
   = note: inside `get_or_init` at tests/genmc/pass/atomics/cas_failure_ord_racy_init.rs:LL:CC
note: inside closure
  --> tests/genmc/pass/atomics/cas_failure_ord_racy_init.rs:LL:CC
   |
LL |             get_or_init(1);
   |             ^^^^^^^^^^^^^^
   = note: inside `<std::boxed::Box<{closure@tests/genmc/pass/atomics/cas_failure_ord_racy_init.rs:LL:CC}> as std::ops::FnOnce<()>>::call_once` at RUSTLIB/alloc/src/boxed.rs:LL:CC
note: inside `genmc::spawn_pthread_closure::thread_func::<{closure@tests/genmc/pass/atomics/cas_failure_ord_racy_init.rs:LL:CC}>`
  --> tests/genmc/pass/atomics/../../../utils/genmc.rs:LL:CC
   |
LL |         f();
   |         ^^^

Verification complete with 8 executions. No errors found.
Number of complete executions explored: 2
Number of blocked executions seen: 6
Verification took [TIME]s.
